### DISCUSSION TITLE
Return response from tRPC middlware

### DISCUSSION THREAD

**Discussion by wanschi** [trpc/trpc#3929] [Q&A]
Hi guys,

we're using tRPC in a team of 5 and are enjoying the advantages so far. We're currently building a demo mode for our app. Requests from the client with demo mode enabled should always get a static response from the tRPC server. Our idea was to build a tRPC middleware, which should return the static response if the demo mode was detected within the middleware. In express that would be possible because we have access to the `res` object within in the middleware. In tRPC I'm forced to return `next()` as return value of the middleware and have no possibility to interact with the `res` object. 

Maybe I'm overseeing something or you have another idea.

Thank you

**Comment by unshame** [comment]
I have a similar question, but it's about a redis cache middleware. I want to have something like this.

```typescript
const getSomething = procedure
    .use(cacheMiddleware('somePrefixKey'))
    .query(...);
```

and cacheMiddleware is something like

```typescript
function cacheMiddleware(prefixKey) {
     return middleware(async ({ input, next }) => {
         const cacheKey = `${prefixKey}:${JSON.stringify(input)}`;
         const cache = await redis.get(cacheKey);

        if (cache) {
              return { ok: true, data: cache } // ts error - missing marker property
        }

        const result = await next();

        cache.set(cacheKey, result.data);

        return result;
     })
}
```

I found it's possible to return

```typescript
{
                ok: true,
                data,
                marker: 'middlewareMarker' as 'middlewareMarker' & {
                    __brand: 'middlewareMarker';
                },
            }
```

from middleware to satisfy typescript (the return type is branded), but that seems like a bad idea, especially considering the comment next to the marker property states

> All middlewares should pass through their next()'s output. Requiring this marker makes sure that can't be forgotten at compile-time.

I think for your use case this is fine, since it's for a demo mode. But I'm wondering how bad of an idea this is for a caching middleware. Is not calling next but returning a result a potential footgun?

**Comment by ivanbarlog** [comment]
Hi @wanschi @unshame,

I have a similar problem which I am trying to solve but unfortunately your solution does not work for me.

I need to set some generic response - either error or 200OK with empty body - and also I would like to change headers within my middleware.

My use case is as follows:

1. request is processed in my `createContext()` function where I parse cookie which I get from `req` parameter, verify the user has valid session and get some user data like `userId` from the database.  I am also setting header through `resHeaders` parameter.
2. I return `req`, adjusted `resHeaders` and `userId` - when the cookie is not valid or it has expired I am returning `null` instead of the real `userId`
3. I have created `protectedRoute` middleware where I am checking if the `userId` is `null` or not - if it is not `null` I just return the `next()` but if it IS `null`, I want to return response

In the 3rd step when `userId` is `null` I've tried to return error and also I've tried to return some data as 200 OK response which works but it seems that whatever I have set in the `createContext()` in `resHeaders` seems to be ignored for some reason.

I simply want to invalidate cookie which was expired by setting `set-cookie` header but this is not possible in the middleware and the one I've set in the `createContext` seems to be ignored.

Do you have any experience with this?
